lib/cdo/firehose.rb

```diff
@@ -18,31 +19,109 @@
 #       data_json: "{\"key\":\"value\"}"    # OPTIONAL
 #     }
 #   )
+# The `data` in the records which are sent to Firehose will be JSON encoded, so make sure your Redshift "Copy options"
+# have `json 'auto'`. This will tell Redshift to parse the data as a JSON blob.
+#
+# All records uploaded to Firehose will have some common attributes added to them which you can optionally tell Firehose
+# top copy.
+#   `created_at` timestamp - The timestamp of when this record was created.
+#   `environment` varchar(128) - The Ruby RACK_ENV which produced this record e.g. 'Production', 'Test'
+#   `device` varchar(1024) - JSON blob with information about the source of the record
+#    e.g. "server-side", {"useragent: ..."}
 
-STREAM_NAME = 'analysis-events'.freeze
+ANALYSIS_EVENTS_STREAM_NAME = 'analysis-events'.freeze
+# This Firehose stream is used for collecting data about string:url associations.
+I18N_STRING_TRACKING_EVENTS_STREAM_NAME = 'i18n-string-tracking-events'.freeze
+
+# Experiment name for using the Firehose put_record_batch API
+FIREHOSE_PUT_RECORD_BATCH_DCDO_KEY = 'firehose_put_record_batch'.freeze
 
 class FirehoseClient
   include Singleton
 
   REGION = 'us-east-1'.freeze
+  # Max number of records Firehose allows to be pushed in a batch.
+  FIREHOSE_PUT_MAX_RECORDS = 500
 
   # Initializes the @firehose to an AWS Firehose client.
   def initialize
-    if [:development, :test].include? rack_env
+    @rack_env = rack_env
+    if [:development, :test].include? @rack_env
       return
     end
     @firehose = Aws::Firehose::Client.new(region: REGION)
   end
 
+  # Posts records in batches to the given AWS Kinesis Firehose stream.
+  # @param stream_name [string] The Kinesis stream to send the data to.
+  # @param datas [array[hash]] The list of data to send to the stream.
+  def put_record_batch(stream_name, datas)
+    return if datas.nil? || datas.empty?
+    return unless Gatekeeper.allows('firehose', default: true)
+
+    # convert the given data into the format Firehose expects
+    records = []
+    datas.each do |data|
+      records << create_record_from_data(add_common_values(data))
+    end
```

Comment:
Use [`#map`](https://ruby-doc.org/core/Array.html#method-i-map):
```suggestion
    records = datas.map do |data|
      create_record_from_data(add_common_values(data))
    end
```